### ISSUE TITLE
Repackage node2nix generated packages using buildNpmPackage

### ISSUE THREAD

**Issue by ehllie** [NixOS/nixpkgs#229475]
## Issue description
Currently a vast majority of NPM packages available in Nixpkgs are generates using node2nix. It generates a nix expression form the [`node-packages.json` file](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/node-packages.json). However it's very hard to maintain. Any time a new package is added to that list or updated, node2nix has to update every single package listed in `node-packages.json` using a [code-gen script](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/generate.sh). That forces PRs that touch that part of the ecosystem to rerun the code-gen script each time only other PR is merged. This is further exacerbated by the time the script takes to run. As of writing this post, there are 408 packages in `node-packages.json`, and it takes ~4h to run the generation script.

## Proposed solution
Nowadays an alternative tool, [`buildNpmPackage`](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/javascript.section.md#buildnpmpackage-javascript-buildnpmpackage), is available, but plenty of the packages already listed there were added there before its creation. Not all of the packages can be build using the new tool, but I think it would be worthwhile to review how many of them could be repackaged that way. That should at the very least make the time of running the code-gen script shorter.

As such, I've listed all the packages that are build using `node2nix`. My intention is this issue could be used to collaborate on the repackaging efforts. If any package gets successfully repackaged, or determined to not be possible to be built using `builtNpmPackage`, I'd appreciate having that mentioned here so that I can update the list and move that package to an appropriate location. 

<details>
<summary>Not checked</summary>

- @angular/cli
- @antfu/ni
- @antora/cli
- @antora/site-generator-default
- @astrojs/language-server
- @bitwarden/cli
- @commitlint/cli
- @commitlint/config-conventional
- @emacs-eask/cli
- @forge/cli
- @google/clasp
- @medable/mdctl-cli
- @microsoft/rush
- @nerdwallet/shepherd
- @nestjs/cli
- @shopify/cli
- @squoosh/cli
- @tailwindcss/aspect-ratio
- @tailwindcss/forms
- @tailwindcss/language-server
- @tailwindcss/line-clamp
- @tailwindcss/typography
- @uppy/companion
- @vue/cli
- @webassemblyjs/cli
- @webassemblyjs/repl
- @webassemblyjs/wasm-strip
- @webassemblyjs/wasm-text-gen
- @webassemblyjs/wast-refmt
- alex
- alloy
- antennas
- asar
- audiosprite
- autoprefixer
- auto-changelog
- aws-azure-login
- aws-cdk
- awesome-lint
- balanceofsatoshis
- bash-language-server
- bibtex-tidy
- bower
- bower2nix
- browserify
- browser-sync
- btc-rpc-explorer
- castnow
- carbon-now-cli
- carto
- cdk8s-cli
- cdktf-cli
- clean-css-cli
- clipboard-cli
- clubhouse-cli
- coc-clangd
- coc-cmake
- coc-css
- coc-diagnostic
- coc-docker
- coc-emmet
- coc-eslint
- coc-explorer
- coc-flutter
- coc-git
- coc-go
- coc-haxe
- coc-highlight
- coc-html
- coc-imselect
- coc-java
- coc-jest
- coc-json
- coc-lists
- coc-ltex
- coc-markdownlint
- coc-metals
- coc-pairs
- coc-prettier
- coc-pyright
- coc-python
- coc-r-lsp
- coc-rls
- coc-rust-analyzer
- coc-sh
- coc-smartf
- coc-snippets
- coc-solargraph
- coc-spell-checker
- coc-stylelint
- coc-sumneko-lua
- coc-sqlfluff
- coc-tabnine
- coc-texlab
- coc-toml
- coc-tslint
- coc-tslint-plugin
- coc-tsserver
- coc-ultisnips
- coc-vetur
- coc-vimlsp
- coc-vimtex
- coc-wxml
- coc-yaml
- coc-yank
- code-theme-converter
- coffee-script
- coinmon
- concurrently
- configurable-http-proxy
- conventional-changelog-cli
- cordova
- cpy-cli
- create-cycle-app
- create-react-app
- create-react-native-app
- cspell
- csslint
- dat
- degit
- dhcp
- diagnostic-languageserver
- diff2html-cli
- dockerfile-language-server-nodejs
- elasticdump
- @electron-forge/cli
- eas-cli
- elm-oracle
- elm-test
- emoj
- emojione
- epgstation
- epgstation-client
- escape-string-regexp
- eslint
- eslint_d
- esy
- expo-cli
- fast-cli
- fauna-shell
- firebase-tools
- fixjson
- fkill-cli
- fleek-cli
- flood
- forever
- fx
- ganache
- gatsby-cli
- generator-code
- get-graphql-schema
- git-run
- git-ssb
- git-standup
- @gitbeaker/cli
- gitmoji-cli
- glob
- gramma
- grammarly-languageserver
- graphql
- graphql-cli
- graphql-language-service-cli
- graphqurl
- grunt-cli
- makam
- meshcommander
- gqlint
- gtop
- gulp
- gulp-cli
- he
- html-minifier
- htmlhint
- http-server
- hsd
- hs-airdrop
- hs-client
- hueadm
- hyperpotamus
- ijavascript
- inliner
- imapnotify
- immich
- indium
- insect
- intelephense
- ionic
- jake
- javascript-typescript-langserver
- joplin
- js-beautify
- js-yaml
- jsdoc
- jshint
- json
- json-diff
- json-refs
- json-server
- jsonlint
- jsonplaceholder
- kaput-cli
- katex
- karma
- keyoxide
- lcov-result-merger
- leetcode-cli
- vsc-leetcode-cli
- lerna
- less
- less-plugin-clean-css
- live-server
- livedown
- localtunnel
- lodash
- lua-fmt
- lv_font_conv
- madoko
- manta
- markdownlint-cli
- markdownlint-cli2
- markdown-link-check
- markdown-preview-nvim
- mastodon-bot
- mathjax
- meat
- @mermaid-js/mermaid-cli
- mocha
- multi-file-swagger
- musescore-downloader
- near-cli
- neovim
- nijs
- node-gyp
- node-gyp-build
- node-inspector
- node-pre-gyp
- node-red
- node2nix
- nodemon
- np
- npm
- npm-check-updates
- npm-merge-driver
- nrm
- ocaml-language-server
- orval
- parcel-bundler
- parcel
- parsoid
- patch-package
- peerflix
- peerflix-server
- pkg
- pm2
- pnpm
- poor-mans-t-sql-formatter-cli
- postcss
- postcss-cli
- prebuild-install
- prettier
- prettier_d_slim
- prettier-plugin-toml
- prisma
- @prisma/language-server
- pscid
- pulp
- purescript-language-server
- purescript-psa
- purs-tidy
- purty
- pxder
- pyright
- quicktype
- react-native-cli
- react-static
- react-tools
- readability-cli
- redoc-cli
- remod-cli
- reveal.js
- reveal-md
- rimraf
- rollup
- rust-analyzer-build-deps
- rtlcss
- s3http
- sass
- semver
- serve
- serverless
- shout
- sloc
- smartdc
- snyk
- socket.io
- speed-test
- sql-formatter
- ssb-server
- stackdriver-statsd-backend
- stf
- stylelint
- surge
- svelte-check
- svelte-language-server
- svgo
- swagger
- tailwindcss
- tedicross
- teck-programmer
- tern
- terser
- textlint
- textlint-plugin-latex
- textlint-rule-abbr-within-parentheses
- textlint-rule-alex
- textlint-rule-common-misspellings
- textlint-rule-diacritics
- textlint-rule-en-max-word-count
- textlint-rule-max-comma
- textlint-rule-no-start-duplicated-conjunction
- textlint-rule-period-in-list-item
- textlint-rule-stop-words
- textlint-rule-terminology
- textlint-rule-unexpanded-acronym
- textlint-rule-write-good
- thelounge
- thelounge-plugin-closepms
- thelounge-plugin-giphy
- thelounge-plugin-shortcuts
- thelounge-theme-abyss
- thelounge-theme-amoled
- thelounge-theme-amoled-sourcecodepro
- thelounge-theme-bdefault
- thelounge-theme-bmorning
- thelounge-theme-chord
- thelounge-theme-classic
- thelounge-theme-common
- thelounge-theme-crypto
- thelounge-theme-discordapp
- thelounge-theme-dracula
- thelounge-theme-dracula-official
- thelounge-theme-flat-blue
- thelounge-theme-flat-dark
- thelounge-theme-gruvbox
- thelounge-theme-hexified
- thelounge-theme-ion
- thelounge-theme-light
- thelounge-theme-midnight
- thelounge-theme-mininapse
- thelounge-theme-monokai-console
- thelounge-theme-mortified
- thelounge-theme-neuron-fork
- thelounge-theme-new-morning
- thelounge-theme-new-morning-compact
- thelounge-theme-nologo
- thelounge-theme-nord
- thelounge-theme-onedark
- thelounge-theme-purplenight
- thelounge-theme-scoutlink
- thelounge-theme-seraphimrp
- thelounge-theme-solarized
- thelounge-theme-solarized-fork-monospace
- thelounge-theme-zenburn
- thelounge-theme-zenburn-monospace
- thelounge-theme-zenburn-sourcecodepro
- three
- tiddlywiki
- titanium
- triton
- tsun
- ts-node
- ttf2eot
- typescript
- typescript-language-server
- uglify-js
- undollar
- ungit
- unified-language-server
- vega-cli
- vega-lite
- vercel
- vim-language-server
- vls
- vscode-css-languageserver-bin
- vscode-html-languageserver-bin
- vscode-json-languageserver
- vscode-json-languageserver-bin
- vscode-langservers-extracted
- vue-cli
- vue-language-server
- wavedrom-cli
- web-ext
- webpack
- webpack-cli
- webpack-dev-server
- copy-webpack-plugin
- webtorrent-cli
- @withgraphite/graphite-cli
- wrangler
- wring
- write-good
- @yaegassy/coc-nginx
- yaml-language-server
- yalc
- yarn
- yo
- @zwave-js/server
- zx

</details>

<details>
<summary>Repackaged</summary>

None yet

</details>

<details>
<summary>Can't be repackaged</summary>

None yet

</details>

I've discussed this idea with @SuperSandro2000 in another PR briefly, but I'd appreciate other people responsible for maintaining the node ecosystem to weigh in as well. @NixOS/node would this be the right way to go about this? Where should the repackaged packages be put? Is doing this worthwhile?

**Comment by winterqt** [comment]
Feel free to move packages out of the set, this is a net good. Not really sure how/if we want to add aliases for the old names, though. (cc @lilyinstarlight)

As a rule of thumb, *more or less* anything with a package-lock.json can be packaged, but some are impossible due to npm cursedness.

Feel free to request my review directly on packages, I'm happy to land things quickly and work with you on this :)

**Comment by natsukium** [comment]
Just the other day, I came across a PR to change from nodePackages to a standalone build. (#228854)
In this PR, an alias was added to `override.nix` to keep the nodePackages name. https://github.com/NixOS/nixpkgs/pull/228854/commits/b4b15b69cb9afe1cd6b57c43ab031b49cbfce971

I would also like to help migrate the package.

**Comment by winterqt** [comment]
Ah, forgot about the overrides file. That looks good, do that. Again, please request my review on these, I'm more than happy to review and get them landed.

Please do keep in mind that we expect you to maintain a package if you add it, so only migrate packages that you are okay with maintaining and testing.

**Comment by lilyinstarlight** [comment]
@winterqt Should we make a `node-aliases.nix` kinda like Python has now (either in `top-level` or still in the `node-packages` dir)?

That way they are easy to find and turn into throws later (as for other aliases) and we can just make the node-packages `default.nix` optionally include it on `config.allowAliases`

**Comment by winterqt** [comment]
Great catch, yes, let's do it.

**Comment by ehllie** [comment]
I've gone over a couple of the packages so far, and it seems to be not as straightforward as I initially thought it would be 😛

- @angular/cli and @commitlint/cli use yarn, and do not have a `package-lock.json`
- @antfu/ni and @astrojs/language-server use pnpm, and likewise have no `package-lock.json`
- @nestjs/cli and @squoosh/cli fail due to issue with building `node-gyp`, but those seem to be addressed in #206477, I'll investigate if that PR makes them possible to build

**Comment by natsukium** [comment]
For a start, I have categorized the packages around the ones I use.
I know it will be hard to update the list, but please check it out.

<details>
<summary>Repackage candidate (has package-lock.json)</summary>

- @google/clasp
- @tailwindcss/language-server
- @tailwindcss/line-clamp
- @tailwindcss/typography
- create-react-app
- dockerfile-language-server-nodejs
- markdownlint-cli
- npm
- pyright
- tailwindcss
- textlint
- textlint-rule-diacritics
- textlint-rule-stop-words
- textlint-rule-terminology
- wrangler
</details>

<details>
<summary>Repackaged</summary>

- @bitwarden/cli #218923 
</details>

<details>
<summary>Can't be repackaged</summary>

#### `No package-lock.json`
##### with `yarn`
- bash-language-server
- markdown-preview-nvim
- prettier
- textlint-plugin-latex
- textlint-rule-abbr-within-parentheses
- textlint-plugin-latex
- textlint-rule-en-max-word-count
- textlint-rule-max-comma
- textlint-rule-no-start-duplicated-conjunction
- textlint-rule-period-in-list-item
- textlint-rule-unexpanded-acronym
- textlint-rule-write-good
- yarn

##### with `pnpm`
- conventional-changelog-cli
- cspell
- grammarly-languageserver
- pnpm
- vercel

##### with `npm` but no package-lock.json
- @tailwindcss/aspect-ratio
- @tailwindcss/forms
- eslint
- markdownlint-cli2
- textlint-rule-common-misspellings
</details>

**Comment by SuperSandro2000** [comment]
I think we should start with everything that is easy to get done and has a top-level alias. Maybe that is even enough to get nodePackages into shape and we can save ourselves time fixing the really hard and weird cases.

**Comment by ehllie** [comment]
@natsukium thanks a lot for compiling these packages. I've edited the original post

**Comment by lilyinstarlight** [comment]
I've added aliases for `nodePackages` in #229639

**Comment by malob** [comment]
Just want to note that a missing `package-lock.json` in the source isn't necessarily a blocker for using `buildNpmPackage`.

The `package-lock.json` file can be generated and committed to `nixpkgs`. This is what I had to do to add `github-copilot-cli`: https://github.com/NixOS/nixpkgs/blob/1e237b7c3c0d544a0b1efacf22849a0ff261f7a6/pkgs/tools/misc/github-copilot-cli/default.nix#L14-L16

**Comment by Mic92** [comment]
Can we please node add 5000 lines of generated packages.json locks per node package?
https://github.com/NixOS/nixpkgs/pull/231615/files
Otherwise we would end up with 19.1M lines of code just for all nodes packages. This would significantly increase the size of nixpkgs. I am strongly against that.

**Comment by Mic92** [comment]
Last time I checked node2nix was doing a lot of the fetching in serial. However one could speed it up by pulling with 32-threads instead of just one (7min instead of 4hours based on my napkin math)

**Comment by natsukium** [comment]
Indeed, in the example above, it was undesirable to commit the 5,000 lines of package-lock.json.
On the other hand, if the upstream keeps the package-lock.json, we don't have to add it ourselves and can benefit from a much smaller number of lines in the `node-package.nix`.

Also, even if we could fetch in parallel, my laptop (macbook pro) only has eight threads, so it is unlikely to be that fast. (although it is enough compared to 4 hours).

**Comment by RaitoBezarius** [comment]
Can we get statistics / figures on what it would add to nixpkgs before anything?

**Comment by winterqt** [comment]
FWIW @Mic92, not every package requires that we add a `package-lock.json` in-tree, only some do. I'd argue that we should migrate all the ones that *don't* require us to do that first, and then re-evaluate.

**Comment by RaitoBezarius** [comment]
Apologies for my confusion, I didn't understand it was about adding only those who had package-lock.json in tree. :+1: for the approach!

**Comment by Mic92** [comment]
I am ok with this approach when we do this for packages that provide a `package-lock.json`

**Comment by SuperSandro2000** [comment]
We can also upload the package-lock.json to somewhere with buildNpmPackage.

> Last time I checked node2nix was doing a lot of the fetching in serial. However one could speed it up by pulling with 32-threads instead of just one (7min instead of 4hours based on my napkin math)

That doesn't fix that it is broken in other ways and confuses dev and prod dependencies.

**Comment by ehllie** [comment]
> We can also upload the package-lock.json to somewhere with buildNpmPackage.

That's definitely one way to do it, but we'd need to decide on a hosting platform for said lock files. It's probably better not to leave the choice up to the individual contributor as that would possibly make maintenance even harder than it is now, with the upstream fetcher endpoints going offline and needing to be updated. I'm also not sure if we wouldn't run into IFD issues by doing that.

But that seems to be a general issue with monorepo approach that Nixpkgs has. Any user of the repository needs to clone the entire repository regardless of how much of it they will actually use. It's obviously not cloning the binaries, but Nixpkgs is continuously growing, and I don't know if that approach is sustainable long term. However that question is beyond the scope of this issue and should probably be addressed in an RFC. 

I feel like sticking to packages with pre-existing `package-lock.json` files and making `node2nix` make requests in a concurrent manner like @Mic92 suggested is the best course of action we can take atm.

As a side note, I apologise for starting the issue and then disappearing out of the blue. I had just undergone surgery and didn't have much time in the days leading up to it. I should be able to contribute a bit more now that I'm stuck in a hospital bed however ^^

**Comment by SuperSandro2000** [comment]
> But that seems to be a general issue with monorepo approach that Nixpkgs has. Any user of the repository needs to clone the entire repository regardless of how much of it they will actually use. It's obviously not cloning the binaries, but Nixpkgs is continuously growing, and I don't know if that approach is sustainable long term. However that question is beyond the scope of this issue and should probably be addressed in an RFC.

One of the biggest offender to repo size is nodePackages because the file is just so big.

**Comment by Lord-Valen** [comment]
With `yarn`:
+ yaml-language-server
+ typescript-language-server

With `rush` (thereby `pnpm`):
+ rush

**Comment by pbsds** [comment]
> We can also upload the package-lock.json to somewhere with buildNpmPackage.

Perhaps a separate github repo? We could put each lock file into a separate branch/tag, to avoid downloading unrelated locks. Upon request a bot could fetch a package source, `npm install --package-lock-only`, and commit it to a new branch, then reply with download url and `fetchurl` hash.

**Comment by ehllie** [comment]
> > We can also upload the package-lock.json to somewhere with buildNpmPackage.
> 
> That's definitely one way to do it, but we'd need to decide on a hosting platform for said lock files. It's probably better not to leave the choice up to the individual contributor as that would possibly make maintenance even harder than it is now, with the upstream fetcher endpoints going offline and needing to be updated. I'm also not sure if we wouldn't run into IFD issues by doing that.
> 
> But that seems to be a general issue with monorepo approach that Nixpkgs has. Any user of the repository needs to clone the entire repository regardless of how much of it they will actually use. It's obviously not cloning the binaries, but Nixpkgs is continuously growing, and I don't know if that approach is sustainable long term. However that question is beyond the scope of this issue and should probably be addressed in an RFC.
> 
> I feel like sticking to packages with pre-existing `package-lock.json` files and making `node2nix` make requests in a concurrent manner like @Mic92 suggested is the best course of action we can take atm.
> 
> As a side note, I apologise for starting the issue and then disappearing out of the blue. I had just undergone surgery and didn't have much time in the days leading up to it. I should be able to contribute a bit more now that I'm stuck in a hospital bed however ^^

My idea with splitting up `nixpkgs`, is then we could also just split off any package that needs to have its `package-lock.json` generated into a separate repository, that way it won't affect the size of the actual `nixpkgs` vsc tree, and won't be included in the tarball download unless the they are actually being built. The same could be done with the massive package set in `node-packages.json` for the packages that can't use `buildNpmPackage`, where each of the packages would get it's own repository with the code-gen script, and would be updated on a per need basis, or periodically with a GitHub action. However, I think that would require nixpkgs to further adopt flakes in general. @winterqt @SuperSandro2000 @lilyinstarlight do you recon that would be worth making a separate issue about? Because I think I's worthwhile continuing to repackage things the way I suggested in this thread, but my new idea would probably affect more of `nixpkgs` as a whole, and be taking a somewhat different approach.

**Comment by SuperSandro2000** [comment]
Relying on flakes for nixpkgs is not something that should happen. Also anything that triggers IFD must be avoided. I think the package-lock.json is in a unique position that it is not read by nix directly, so we can move it to somewhere else. For ``node-packages.json`` that does not work. Also anything requiring a RFC should be avoided as it will delay things for multiple months.

> where each of the packages would get it's own repository with the code-gen script

That would create an insane amount of unmanageable repositories. We can also stuff that into repo with subdirectories.

**Comment by stasjok** [comment]
> With `yarn`:
> 
>     * yaml-language-server
> 

Formatting doesn't work in current version of `nodePackages.yaml-language-server`. I think this is because prettier was updated to version 3.0.0 recently. It is added to yaml-language-server manually here:

https://github.com/NixOS/nixpkgs/blob/7c70396c1de0d0e588ff226b795d54459b135227/pkgs/development/node-packages/overrides.nix#L609-L615

but upstream uses version 2.6.7. I think the only way to fix it is to repackage it with `buildNpmPackage` so that it doesn't depend on `nodePackages.prettier`, but to a locked version.

**Comment by dotlambda** [comment]
The following packages use their own `node-packages.nix` and should also be repackaged if possible:

- [ ] base16-builder (@garbas)
- [ ] elmPackages
- [ ] ethercalc (@iblech)
- [ ] fx_cast_bridge (@dtzWill @kevincox @PedroHLC)
- [ ] n8n (@freezeboy @K900)
- [ ] netlify-cli (@roberth @panda2134)
- [ ] karma (@NukaDuka)
- [ ] ldgallery (@pacien)
- [ ] mx-puppet-discord (@arcnmx @expipiplus1)
- [ ] onlykey (@kalbasit)
- [ ] playwright-test (@teto @kalekseev)
- [ ] polaris (@pbsds)
- [ ] psitransfer https://github.com/NixOS/nixpkgs/pull/245426
- [ ] whitebophir (@iblech)

**Comment by K900** [comment]
I've had some issues with n8n, a look from someone better versed in the JS abyss would be appreciated.

**Comment by marsam** [comment]
>netlify-cli

I have a wip. It builds, but the test is failing https://github.com/NixOS/nixpkgs/pull/243664

**Comment by panda2134** [comment]
> #243664

What OS/platform are you on? AFAIK the tests are failing on macOS due to the sandboxing mechanism.

**Comment by dotlambda** [comment]
> I've had some issues with n8n, a look from someone better versed in the JS abyss would be appreciated.

Looks like n8n uses pnpm, so we can't use `buildNpmPackage` for now.

**Comment by dotlambda** [comment]
Please review https://github.com/NixOS/nixpkgs/pull/285152 which uses `fetchYarnDeps` for a package that uses lerna.

**Comment by tomodachi94** [comment]
I've made a script to allow one to quickly assess where "Not checked" entries should go. It's a bit hacky, but it seems to work pretty well:

```sh
#!/usr/bin/env sh

while IFS= read -r line; do
    src="$(sudo "$(which nix-build)" --quiet -A "nodePackages.$line".src 2>/dev/null || echo /dev/null)"

    if [ -d "$src" ]; then
        if find "$src" -name "package-lock.json" -print -quit | grep -q .; then
            echo "package-lock.json: $line"
        elif find "$src" -name "yarn.lock" -print -quit | grep -q .; then
            echo "yarn.lock: $line"
        elif find "$src" -name "pnpm-lock.yaml" -print -quit | grep -q .; then
            echo "pnpm-lock.yaml: $line"
        else
            echo "unrepackagable: $line"
        fi
    else
        echo "source not found: $line"
    fi
done
```
(Feed it a newline-delimited list without leading whitespace)

**Comment by tomodachi94** [comment]
I have moved the list to https://github.com/NixOS/nixpkgs/projects/83 to hopefully make manipulating it easier.

**Comment by tomodachi94** [comment]
To repackage `thelounge` plugins, we will need to reimplement the `theLoungePlugins` attrset:

https://github.com/NixOS/nixpkgs/blob/02289590e9ff14e8e3aebeedd54eddccebef61bc/pkgs/top-level/all-packages.nix#L12798-L12806

**Comment by pyrox0** [comment]
> To repackage `thelounge` plugins, we will need to reimplement the `theLoungePlugins` attrset:
> 
> https://github.com/NixOS/nixpkgs/blob/02289590e9ff14e8e3aebeedd54eddccebef61bc/pkgs/top-level/all-packages.nix#L12798-L12806

I think it should be implemented as two separate attrsets. one called `theLoungePlugins` and one called `theLoungeThemes`. the previous one seemed to be `theLoungePlugins.plugins` and `theLoungePlugins.themes`, which feels very clunky. Perhaps just 2 separate folders in thelounge's package folder with nix files for each? should be the simplest solution, hell if they all use the same build system, could we do a common.nix approach?

**Comment by tomodachi94** [comment]
I don't think a `common.nix` approach will be that simple. The packages are packaged using a smattering of different tooling:

https://github.com/orgs/NixOS/projects/83/views/2?filterQuery=thelounge-+-status%3A%22Not+checked%22+

**Comment by pyrox0** [comment]
Hello to anyone still following this issue. I've made some big progress in the last few weeks, and the node-packages.nix file is now just a "mere" 1.25Mib and 35558 LoC(meaning it can be displayed in the Github WebUI now!). however, from what I can tell, most of the remaining 63 packages are either difficult(as in, I've tried before and failed because they suck) or impossible to package atm(mostly because of no lockfiles being available for those packages)

Therefore, while this issue is nowhere near done, it is definitely in a much better place than where it was even just 2 months ago(in fact, it's dropped nearly 1MiB since Oct 1st!). 

Along with that, 66/93 packages removed from nodePackages during that time were added into the main nixpkgs set, with no additional bloaty lockfiles to speak of. I'm very proud of the work I've done, but of course there is always more work to do. If anyone would like to help me package the last few difficult ones with lockfiles(See https://github.com/orgs/NixOS/projects/83/views/1), that would be great!

**Comment by mweinelt** [comment]
Don't beat yourself up about the remaining packages. That's shitty work nobody should spend any time on in the first place. I suggest dropping the packages that are not leaf packages and then going through the rest on a case by case basis.

It's been years since we knew nodePackages was a bad idea, we need it to end without burning you out.
